doc/sr_winnow.1.rst
==========
 SR_Winnow 
==========

-----------------------------------------------------------
winnows messages, suppressing redundant product notification
-----------------------------------------------------------

:Manual section: 1 
:Date: @Date@
:Version: @Version@
:Manual group: MetPX-Sarracenia

SYNOPSIS
========

**sr_winnow** configfile foreground|start|stop|restart|reload|status

DESCRIPTION
===========

**sr_winnow** is a program that Subscribes to file notifications, 
and reposts the notifications suppressing the redundant files.

To understand what the **sum** is, the notification protocol is
defined here `sr_post(7) <sr_post.7.html>`_

**sr_winnow** connects to a *broker* (often the same as the posting broker)
and subscribes to the notifications of interest. On reception if a notification,
it uses its **sum** to verify if the file is already in its cache... 
if it is, the notification is ignored, if not, the **sum** is added to the cache
and the notification is posted.  

**sr_winnow** can be used to trim messages from `sr_post(1) <sr_post.1.html>`_,
`sr_poll(1) <sr_poll.1.html>`_  or `sr_watch(1) <sr_watch.1.html>`_  etc...

The **sr_winnow** command takes two argument: a configuration file described below,
followed by an action start|stop|restart|reload|status... (self described).

The **foreground** action is different. It would be used when building a configuration
or debugging things. It is used when the user wants to run the program and its configfile 
interactively...   The **foreground** instance is not concerned by other actions. 
The user would stop using the **foreground** instance by simply pressing <ctrl-c> on linux 
or use other means to kill its process.

CONFIGURATION
=============

Options are placed in the configuration file, one per line, of the form: 

**option <value>** 

Comment lines begins with **#**. 
Empty lines are skipped.
For example::

  **debug true**

would be a demonstration of setting the option to enable more verbose logging.
The configuration default for all sr_* commands is stored in 
the ~/.config/sarra/default.conf file, and while the name given on the command 
line may be a file name specified as a relative or absolute path, sr_winnow 
will also look in the ~/.config/sarra/winnow directory for a file 
named *config.conf*  The configuration in specific file always overrides
the default, and the command line overrides any configuration file.

ACTIVE/PASSIVE OPTIONS
----------------------

**sr_winnow** can be used on a single server.
Idealy, to make it more robust, you would run it on clustered brokers.
An high availability software presents a **vip** (virtual ip) on the active
server. Should the server go down, the **vip** is moved on another server.
Both servers would run **sr_winnow**. It is for that reason that the 
following options were implemented:

 - **vip          <string>          (mandatory)** 
 - **interface    <string>          (mandatory)**

In the case, where you run only one **sr_winnow** on one server,
on linux, you would set these options to :

**vip 127.0.0.1**
**interface lo**

In the case of clustered brokers, you would set the options for the 
moving vip.

**vip 153.14.126.3**
**interface eth0**

When **sr_winnow** does not find the vip, it sleeps for 5 seconds and retries.
If it does, it consumes and process a message and than rechecks for the vip.

CREDENTIALS 
-----------

Ther username and password or keys used to access servers are credentials.
For all **sarracenia** programs, the confidential parts of credentials are stored
only in ~/.conf/sarra/credentials.conf.  This includes the broker passwords and settings 
needed by **sr_winnow**.  The format is one entry per line.  Examples:

- **amqp://user1:password1@host/**
- **amqps://user2:password2@host:5671/dev**

In other configuration files or on the command line, the url simply lacks the 
password or key specification.  The url given in the other files is looked 
up in credentials.conf. 


SOURCE NOTIFICATION OPTIONS
---------------------------

First, the program needs to set all the rabbitmq configurations for a source 
broker.  The broker option sets all the credential information to connect 
to the **AMQP** server 

**broker amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**

::

      (default: None and it is mandatory to set it ) 


Once connected to an AMQP broker, the user needs to bind a queue
to exchanges and topics to determine the messages of interest.

QUEUE BINDINGS OPTIONS
----------------------

First, the program needs to set all the rabbitmq configurations for a source broker.
These options define which messages (URL notifications) the program receives:

- **exchange      <name>         (MANDATORY)** 
- **topic_prefix  <name>         (default: v02.post)**
- **subtopic      <amqp pattern> (default: #)**

The **exchange** is mandatory.

If **sr_winnow** is to be used to winnow products from a source 
(**sr_post**, **sr_watch**, **sr_poll**)  then the exchange would
be name 'xs\_'SourceUserName.  SourceUserName is the amqp username the source
uses to announce his products.

topic_prefix is primarily of interest during protocol version transitions,
where one wishes to specify a non-default protocol version of messages to subscribe to. 

To give a correct value to the subtopic, browse the remote server and
write down the directory of interest separated by a dot
as follow:

 **subtopic  directory1.*.subdirectory3.*.subdirectory5.#** 

::

 where:  
       *                replaces a directory name 
       #                stands for the remaining possibilities

The concatenation of the topic_prefix + . + subtopic gives the AMQP topic
One has the choice of filtering using  **topic**  with only AMQP's limited 
wildcarding. 

QUEUE SETTING OPTIONS
---------------------

 - **queue_name   <string>          (default: None)** 
 - **durable      <boolean>         (default: False)** 
 - **expire       <minutes>         (default: None)**
 - **message-ttl  <minutes>         (default: None)**

These options (except for queue_share)  are all AMQP queue attributes.
If a **queue_name** is not provided, it is automatically build by the program.
The name has the form :  q\_'brokerUsername'.sr_winnow.'config_name'
It is easier to have this fix name when it is time to look on the broker
and determine the queue of the program... to see if it is on problem for example.
The program forces the option **queue_share** to True and the option **instances** to 1.
This means, only one instance on one server running (cannot share cache),
but it also means, another instance on another server can access the queue
(when vip fallover another server)

MESSAGE SELECTION OPTIONS
-------------------------

 - **accept        <regexp pattern> (default: False)** 
 - **reject        <regexp pattern> (default: False)** 
 - **on_message            <script> (default: None)** 

One has the choice of filtering using  **subtopic**  with only AMQP's limited 
wildcarding, and/or with the more powerful regular expression based  **accept/reject**  
mechanisms described below.  The difference being that the AMQP filtering is 
applied by the broker itself, saving the notices from being delivered to the 
client at all. The  **accept/reject**  patterns apply to messages sent by the 
broker to the subscriber.  In other words,  **accept/reject**  are client 
side filters, whereas  **subtopic**  is server side filtering.  

It is best practice to use server side filtering to reduce the number of 
announcements sent to the client to a small superset of what is relevant, and 
perform only a fine-tuning with the client side mechanisms, saving bandwidth 
and processing for all.

**sr_winnow** does not check, in the received message, the destination clusters. 
So no message is discarted if without destination, source or other missing attributs.

The user can provide an **on_message** script. When a message is accepted up 
to this level of verification, the **on_message** script is called... with 
the **sr_winnow** class instance as argument.  The script can perform whatever 
you want... if it returns False, the processing of the message will stop 
there. If True, the program will continue processing from there.  


BROKER LOGGING OPTIONS
----------------------

 - **log_exchange     <nane>   (default xlog)**

The state and actions performed with the messages/products of the broker
are logged back to it again through AMQP LOG MESSAGES.  When the broker
pulls products from sources and announces the products on himself, the
**log_exchange** should be set to 'xlog'.  

 
OUTPUT NOTIFICATION OPTIONS
---------------------------

The notifications that are not ignored by **sr_winnow** are reposted.
The options ar

**post_broker amqp{s}://<user>:<pw>@<brokerhost>[:port]/<vhost>**
**post_exchange     <name>         (MANDATORY)** 
**on_post           <script>       (default: None)** 

The **post_broker** defaults to the input broker if not provided.
Just set it to another broker if you want to send the notifications
elsewhere.

The **post_exchange** must be set by the user. This is the exchange under
which the notifications will be posted.

The user can provide an **on_post** script. Just before the message gets
publish to the **post_broker** and under the **post_exchange**, the 
**on_post** script is called... with the **sr_winnow** class instance as argument.
The script can perform whatever you want... if it returns False, the message will not 
bepublished. If True, the program will continue processing from there.  

SEE ALSO
========

`sr_config(7) <sr_config.7.html>`_ - the format of configurations for MetPX-Sarracenia.

`sr_log(7) <sr_log.7.html>`_ - the format of log messages.

`sr_post(1) <sr_post.1.html>`_ - post announcemensts of specific files.

`sr_post(7) <sr_post.7.html>`_ - The format of announcements.

`sr_subscribe(1) <sr_subscribe.1.html>`_ - the download client.

`sr_watch(1) <sr_watch.1.html>`_ - the directory watching daemon.

`dd_subscribe(1) <dd_subscribe.1.html>`_ - the http-only download client.


sarra/sr_winnow.py
#!/usr/bin/python3
#
# This file is part of sarracenia.
# The sarracenia suite is Free and is proudly provided by the Government of Canada
# Copyright (C) Her Majesty The Queen in Right of Canada, Environment Canada, 2008-2015
#
# Questions or bugs report: dps-client@ec.gc.ca
# sarracenia repository: git://git.code.sf.net/p/metpx/git
# Documentation: http://metpx.sourceforge.net/#SarraDocumentation
#
# sr_winnow.py : python3 program allowing to winnow duplicated messages
#                and post the unique and first message in.
#
#
# Code contributed by:
#  Michel Grenier - Shared Services Canada
#  Murray Rennie  - Shared Services Canada
#  Last Changed   : Dec  8 15:22:58 GMT 2015
#  Last Revision  : Jan  8 15:03:11 EST 2016
#
########################################################################
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 2 of the License, or
#  (at your option) any later version.
#
#  This program is distributed in the hope that it will be useful, 
#  but WITHOUT ANY WARRANTY; without even the implied warranty of 
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
#
#

import os,sys,time

try :    
         from sr_amqp           import *
         from sr_consumer       import *
         from sr_instances      import *
         from sr_message        import *
         from sr_poster         import *
except : 
         from sarra.sr_amqp      import *
         from sarra.sr_consumer  import *
         from sarra.sr_instances import *
         from sarra.sr_message   import *
         from sarra.sr_poster    import *

class sr_winnow(sr_instances):

    def __init__(self,config=None,args=None):
        sr_instances.__init__(self,config,args)

        self.cache          = {}
        self.maxEntries     = 12000

    def cache_add(self, key):
        if len(self.cache) >= self.maxEntries: self.cache_clean()
        self.cache[key] = time.time()

    def cache_clean(self):
        temp = [ (item[1],item[0]) for item in self.cache.items() ]
#       self.logger.info("In cache_clean: is it here? 3")
        temp.sort()
#       self.logger.info("In cache_clean: is it here? 2 {0}"  .format(temp[self.maxEntries//2]))
        half,md5 = temp[self.maxEntries//2]
#       self.logger.info("In cache_clean: is it here? half: %s" % (half))
        for item  in self.cache.items():
            if item[1] <= half: 
#               self.logger.info("In cache_clean: delete {0}" .format(item)) 
                r=dict(self.cache)
                del r[item[0]]
                #del self.cache[item[0]]
                self.cache=r

    def cache_clear(self):
        self.cache = {}

    def cache_find(self, key):
        return key in self.cache

    def check(self):

        # When doing winnowing we might switch server...

        self.queue_share = True

        # no queue name allowed

        if self.queue_name == None:
           self.queue_name  = 'q_' + self.broker.username + '.'
           self.queue_name += self.program_name + '.' + self.config_name 

        # we cannot have more than one instance since we 
        # need to work with a single cache.

        if self.nbr_instances != 1 :
           self.logger.warning("Only one instance allowed... set to 1")
           self.nbr_instances = 1

        # exchange must be provided 
        if self.exchange == None:
           self.logger.error("exchange (input) unset...")
           sys.exit(1)

        # by default, post_broker  is the broker
        if self.post_broker == None :
           self.post_broker = self.broker

        # post_exchange must be provided and must be different from exchange
        if self.post_exchange == None or self.post_exchange == self.exchange :
           self.logger.error("post_exchange (output) not properly set...")
           sys.exit(1)

        # bindings should be defined 

        if self.bindings == []  :
           key = self.topic_prefix + '.#'
           self.bindings.append( (self.exchange,key) )
           self.logger.debug("*** BINDINGS %s"% self.bindings)

        # accept/reject
        self.use_pattern          = self.masks != []
        self.accept_unmatch       = True

    def close(self):
        self.consumer.close()
        if self.poster : self.poster.close()

    def connect(self):

        # =============
        # create message
        # =============

        self.msg = sr_message(self.logger)

        # =============
        # consumer
        # =============

        self.consumer          = sr_consumer(self)
        self.msg.log_publisher = self.consumer.publish_back()
        self.msg.log_exchange  = self.log_exchange
        self.msg.user          = self.broker.username

        # =============
        # poster if post_broker different from broker
        # =============

        self.poster = None
        if self.post_broker.geturl() != self.broker.geturl() :
           self.poster            = sr_poster(self)
           self.msg.publisher     = self.poster.publisher

        # =============
        # publisher if post_broker is same as broker
        # =============

        else :
           self.publisher = Publisher(self.consumer.hc)
           self.publisher.build()
           self.msg.publisher = self.publisher
           self.logger.info("Output AMQP  broker(%s) user(%s) vhost(%s)" % \
                           (self.broker.hostname,self.broker.username,self.broker.path) )
        self.msg.pub_exchange  = self.post_exchange
        self.logger.info("Output AMQP  exchange(%s)" % self.msg.pub_exchange )


    def overwrite_defaults(self):

        # default broker : manager

        self.broker      = None
        self.post_broker = None
        if hasattr(self,'manager'):
           self.broker   = self.manager

    def help(self):
        self.logger.info("Usage: %s [OPTIONS] configfile [start|stop|restart|reload|status]\n" % self.program_name )
        self.logger.info("OPTIONS:")
        self.logger.info("-b   <broker>                default manager (if configured)")
        self.logger.info("-e   <exchange>              MANDATORY")
        self.logger.info("-tp  <topic_prefix>          default v02.post")
        self.logger.info("-st  <subtopic>              default #")
        self.logger.info("-pe  <post_exchange>         MANDATORY")
        self.logger.info("DEBUG:")
        self.logger.info("-debug")

    # =============
    # __on_message__
    # =============

    def __on_message__(self):

        # invoke user defined on_message when provided

        if self.on_message : return self.on_message(self)

        return True

    # =============
    # __on_post__ posting of message
    # =============

    def __on_post__(self):

        # invoke on_post when provided

        if self.on_post :
           ok = self.on_post(self)
           if not ok: return ok

        # should always be ok

        ok = self.msg.publish( )

        return ok

    # =============
    # process message  
    # =============

    def process_message(self):

        self.logger.info("Received %s '%s' %s" % (self.msg.topic,self.msg.notice,self.msg.hdrstr))

        #=================================
        # now message is complete : invoke __on_message__
        #=================================

        ok = self.__on_message__()
        if not ok : return ok

        # ========================================
        # cache testing/adding
        # ========================================

        if self.cache_find(self.msg.checksum) :
            self.msg.log_publish(304,'Not modified')
            self.logger.info("Ignored %s" % (self.msg.notice))
            return True

        self.logger.info("Added %s" % (self.msg.notice))
        self.cache_add(self.msg.checksum) 

        # announcing the first and unique message

        self.__on_post__()
        self.msg.log_publish(201,'Published')

        return True


    def run(self):

        # present basic config

        self.logger.info("sr_winnow run")

        # loop/process messages

        self.connect()

        while True :
              try  :
                      #  is it sleeping ?
                      if not self.has_vip() :
                         self.logger.info("sr_winnow is sleeping")
                         time.sleep(5)
                         continue

                      #  consume message
                      ok, self.msg = self.consumer.consume()
                      if not ok : continue

                      #  process message (ok or not... go to the next)
                      ok = self.process_message()

              except:
                      (stype, svalue, tb) = sys.exc_info()
                      self.logger.error("Type: %s, Value: %s,  ..." % (stype, svalue))

    def reload(self):
        self.logger.info("%s reload" % self.program_name)
        self.close()
        self.run()

    def start(self):
        self.logger.info("%s start" % self.program_name)
        self.run()

    def stop(self):
        self.logger.info("%s stop" % self.program_name)
        self.close()
        os._exit(0)
                 
# ===================================
# MAIN
# ===================================

def main():

    action = None
    args   = None
    config = None

    if len(sys.argv) >= 2 : 
       action = sys.argv[-1]

    if len(sys.argv) >= 3 : 
       config = sys.argv[-2]
       args   = sys.argv[1:-2]

    winnow = sr_winnow(config,args)

    if   action == 'foreground' : winnow.foreground_parent()
    elif action == 'reload'     : winnow.reload_parent()
    elif action == 'restart'    : winnow.restart_parent()
    elif action == 'start'      : winnow.start_parent()
    elif action == 'stop'       : winnow.stop_parent()
    elif action == 'status'     : winnow.status_parent()
    else :
           winnow.logger.error("action unknown %s" % action)
           sys.exit(1)

    sys.exit(0)

# =========================================
# direct invocation
# =========================================

if __name__=="__main__":
   main()
